components/camel-weka/src/main/java/org/apache/camel/component/weka/WekaEndpoint.java

```diff
@@ -64,11 +92,245 @@ public boolean isSingleton() {
+            LOG.info("{}", cl);
+        } 
+        
+        // Build a classifier
+        
+        else if (buildSpec != null) {
+            
+            dataset.buildClassifier(buildSpec);
+            
+            // Cross Validate the Model
+            
+            if (crossValidate) {
+                int seed = configuration.getSeed();
+                int folds = configuration.getFolds();
+                dataset.crossValidateModel(folds, seed);
+            } 
+            
+            // Validate the Model using explicit/current instances
+            
+            else {
+                
+                // Use the named data set training
+                if (dsname != null) {
+                    dataset.pop(dsname);
+                }
+                
+                // Train with current instances
+                dataset.evaluateModel();
+            }
+            
+            Classifier cl = dataset.getClassifier();
+            AssertState.notNull(cl, "Model command requires 'load' or 'apply'");
+            LOG.info("{}", cl);
+            
+            Evaluation ev = dataset.getEvaluation();
+            LOG.info("{}", ev.toSummaryString());
+        }
+        
+        // Save the Model
+        
+        if (saveTo != null) {
+            dataset.consumeClassifier(new ModelPersister(saveTo));
+        }
+        
+        return dataset;
+    }
+
+    private Dataset assertDatasetBody(Exchange exchange) throws Exception {
+        
+        Message msg = exchange.getMessage();
+        Object body = msg.getBody();
+        
+        Dataset dataset = msg.getBody(Dataset.class);
+        
+        if (dataset == null) {
+            
+            if (body instanceof Instances) {
+
+                dataset = Dataset.create((Instances) body);
+                
+            } else if (body instanceof GenericFile) {
+                
+                GenericFile<?> file = (GenericFile<?>) body;
+                AssertState.isFalse(file.isDirectory(), "Directory not supported: " + file);
+                String absolutePath = file.getAbsoluteFilePath();
+                dataset = Dataset.create(absolutePath);
+                
+            } else if (body instanceof URL) {
+                
+                URL url = (URL) body;
+                Instances instances = readInternal(url.openStream());
+                dataset = Dataset.create(instances);
+                
+            } else if (body instanceof InputStream) {
+                
+                InputStream input = (InputStream) body;
+                Instances instances = readInternal(input);
+                dataset = Dataset.create(instances);
+            }
+        }
+        
+        AssertState.notNull(dataset, "Cannot obtain dataset from body: " + body);
+        return dataset;
+    }
+
+    // https://github.com/tdiesler/nessus-weka/issues/11
+    private static Instances readInternal(InputStream input) {
```

Comment:
Can the stream not leak? eg need to close it after use

Reply:
Caller resposibility